Riot/Modules/Room/VoiceMessages/VoiceMessagePlaybackController.swift

```diff
@@ -0,0 +1,216 @@
+    
+    private func updateUI() {
+        var details = VoiceMessagePlaybackViewDetails()
+        
+        details.playbackEnabled = (state != .error)
+        details.playing = (state == .playing)
+        details.samples = samples
+        
+        switch state {
+        case .stopped:
+            details.currentTime = durationStringFromTimeInterval(self.duration)
+            details.progress = 0.0
+        default:
+            if let audioPlayer = audioPlayer {
+                details.currentTime = durationStringFromTimeInterval(audioPlayer.currentTime)
+                details.progress = (audioPlayer.duration > 0.0 ? audioPlayer.currentTime / audioPlayer.duration : 0.0)
+            }
+        }
+        details.loading = self.loading
+        
+        playbackView.configureWithDetails(details)
+    }
+        
+    private func loadAttachmentData() {
+        guard let attachment = attachment else {
+            return
+        }
+        
+        self.state = .stopped
+        self.loading = true
+        self.samples = []
+        updateUI()
+        
+        let requiredNumberOfSamples = playbackView.getRequiredNumberOfSamples()
+        
+        cacheManager.loadAttachment(attachment, numberOfSamples: requiredNumberOfSamples) { [weak self] result in
+            guard let self = self else {
+                return
+            }
+            
+            switch result {
+            case .success(let result):
+                guard result.eventIdentifier == attachment.eventId else {
+                    return
+                }
+                
+                // Avoid listening to old audio player delegates if the attachment for this playbackController/cell changes
+                self.audioPlayer?.deregisterDelegate(self)
+                
+                self.audioPlayer = self.mediaServiceProvider.audioPlayerForIdentifier(result.eventIdentifier)
+                self.audioPlayer?.registerDelegate(self)
+                
+                self.loading = false
+                self.urlToLoad = result.url
+                self.duration = result.duration
+                self.samples = result.samples
+                
+                if let audioPlayer = self.audioPlayer {
+                    if audioPlayer.isPlaying {
+                        self.state = .playing
+                    } else if audioPlayer.currentTime > 0 {
+                        self.state = .paused
+                    } else {
+                        self.state = .stopped
+                    }
+                }
+            case .failure:
+                self.state = .error
+            }
+        }
+    }
+    
+    @objc private func updateTheme() {
+        playbackView.update(theme: ThemeService.shared().theme)
+    }
+    
+    private func durationStringFromTimeInterval(_ interval: TimeInterval) -> String {
+        guard interval.isFinite else {
+            return ""
+        }
+        
+        var timeInterval = abs(interval)
```

Comment:
Probably better to use `DateComponentsFormatter` here to allow for any regional format variations?